src/OrchardCore.Modules/OrchardCore.Contents/Drivers/ContentsDriver.cs

```diff
@@ -44,34 +54,59 @@ public override IDisplayResult Display(ContentItem model, IUpdateModel updater)
                         ctx.Shape.Metadata.Alternates.Add($"{stereotype}__ContentsMetadata");
                     }
                 });
+
+                results.Add(contentsMetadataShape);
+
+                var hasViewPermission = await _authorizationService.AuthorizeAsync(context.User, CommonPermissions.ViewContent, contentItem);
```

Comment:
So I guess the advantage of doing all this in the view was that the shapes are built quickly, because this is the front end `DisplayAsync` call and the shapes are not placed.

By moving it here it's more work to render the front end.
(not much)

So is there any advantage to moving all of these out of the views?

For the most part though it looks like you found the issue, and a good solution

Reply:
It's more for the `ActionsMenu:10` shape that sometimes was rendering without having any allowed action(s) to append to it's dropdown.

Reply:
Reverted that code as it was redundant in the view.